### PR TITLE
make killing golem cost no mana

### PR DESCRIPTION
**Pull request by qndel**
Casting golem while there's an active one kills the old one and costs mana. This makes no sense. This change makes the spell only take mana if you summon the golem, killing it is free (Why would you want to kill it? 🤕 + it still requires having enough mana to cast the spell)

### CODE REVIEW

**Comment by qndel** [comment]
> Maybe I'm misunderstanding, but doesn't resummoning restore its life and possibly change it's position. Why should that be free?
Read 🙄

**Comment by StephenCWills** [comment]
I was thinking about this when testing Golem with negative mana. Casting from a scroll or staff would consume the resource just to kill the golem. I guess `UseMana()` handles those cases as well so I very much agree with this change.

Also worth noting that if the player doesn't have the Golem spell and has no more scrolls or charges, they wouldn't even have the option to unsummon their Golem. Best option at that point is to use a Town Portal.

As for why you might want to unsummon him, it's so he doesn't wander off, open doors, and aggro mobs while you're playing Tetris with your inventory. Personally, I really only ever summon him for triple-immunes and then unsummon him afterward.

**Comment by julealgon** [comment]
> Also worth noting that if the player doesn't have the Golem spell and has no more scrolls or charges, they wouldn't even have the option to unsummon their Golem. Best option at that point is to use a Town Portal.

I feel like that makes this change incomplete. We should think about a more robust solution that allows the player to unsummon a minion even if he doesn't have access to the spell anymore.

**Comment by StephenCWills** [comment]
Diablo II had an "Unsummon" skill the player could use to get rid of their summons. I don't know if that's feasible here. If so, we might want to consider whether Guardian can be unsummoned as well.

Alternatively, maybe we could allow casting at slvl 0 to unsummon.

**Comment by Iniquitatis** [comment]
> Diablo II had an "Unsummon" skill the player could use to get rid of their summons. I don't know if that's feasible here. If so, we might want to consider whether Guardian can be unsummoned as well.

Sounds like a proper solution.

> Alternatively, maybe we could allow casting at slvl 0 to unsummon.

Sounds like a dirty hack.

**Comment by StephenCWills** [comment]
> Sounds like a dirty hack.

Agreed. It's also pretty unintuitive.

**Comment by qndel** [comment]
All this stuff goes outside of the scope of the PR, this only fixes a small part of the issue - I just felt having to use mana to unsummon golem was dumb, that's it.

**Comment by FitzRoyX** [comment]
> 
> Read 🙄

Casting golem in beelzebub with an active golem basically refreshes it, but I take it vanilla just kills the existing one without spawning a new one. That's why I'm confused. I don't remember vanilla much and no one gave me the sorc save I asked for to test spells.  Bub indeed added an unsummon hotspell with an X on it if you just want to kill off your summon.

**Comment by AJenbo** [comment]
>  If so, we might want to consider whether Guardian can be unsummoned as well.

You might also wan to kill firewalls as they can eat up all your missiles and take a long time to turn off.

**Comment by nickgavran** [comment]
> > Read 🙄
> 
> Casting golem in beelzebub with an active golem basically refreshes it, but I take it vanilla just kills the existing one without spawning a new one. That's why I'm confused. I don't remember vanilla much and no one gave me the sorc save I asked for to test spells. Bub indeed added an unsummon hotspell with an X on it if you just want to kill off your summon. This could easily just be tied to a hotkey like "delete" if we didn't want to add art.

You really should have read. His original post clearly states that "Casting golem while there's an active one kills the old one and costs mana" - He does not mention that it summons a new one, nor does this PR mention anything about belzebub (which is an entirely different game). You don't have to "take" anything - the description clearly spells out the exact behavior.

**Comment by AJenbo** [comment]
> no one gave me the sorc save I asked for

use debug chats

**Comment by FitzRoyX** [comment]
> > > Read 🙄
> > 
> > 
> > Casting golem in beelzebub with an active golem basically refreshes it, but I take it vanilla just kills the existing one without spawning a new one. That's why I'm confused. I don't remember vanilla much and no one gave me the sorc save I asked for to test spells. Bub indeed added an unsummon hotspell with an X on it if you just want to kill off your summon. This could easily just be tied to a hotkey like "delete" if we didn't want to add art.
> 
> You really should have read. His original post clearly states that "Casting golem while there's an active one kills the old one and costs mana" - He does not mention that it summons a new one, nor does this PR mention anything about belzebub (which is an entirely different game). You don't have to "take" anything - the description clearly spells out the exact behavior.

There's virtually no need to kill an existing one off besides spawning a new one, so why would I assume that vanilla makes you cast twice to refresh a golem? It wasn't even registering in my brain that killing a golem was its own action that cost mana.

**Comment by nickgavran** [comment]
> > > > Read 🙄
> > > 
> > > 
> > > Casting golem in beelzebub with an active golem basically refreshes it, but I take it vanilla just kills the existing one without spawning a new one. That's why I'm confused. I don't remember vanilla much and no one gave me the sorc save I asked for to test spells. Bub indeed added an unsummon hotspell with an X on it if you just want to kill off your summon. This could easily just be tied to a hotkey like "delete" if we didn't want to add art.
> > 
> > 
> > You really should have read. His original post clearly states that "Casting golem while there's an active one kills the old one and costs mana" - He does not mention that it summons a new one, nor does this PR mention anything about belzebub (which is an entirely different game). You don't have to "take" anything - the description clearly spells out the exact behavior.
> 
> There's virtually no need to kill an existing one off besides spawning a new one, so why would I assume that vanilla makes you cast twice to refresh a golem? It wasn't even registering in my brain that killing a golem was its own action that cost mana. Don't act like Bub is so far removed from Diablo, it fixed all of these issues long before this project did, which is why it's all I've been using until now.

Again, you didn't read what neither I, or Qndel wrote. You don't have to 'assume' anything, Qndel's description of the bug/problem clearly states the entire process. I'm not going to post here anymore, this is derailing now.

**Comment by StephenCWills** [comment]
> All this stuff goes outside of the scope of the PR, this only fixes a small part of the issue - I just felt having to use mana to unsummon golem was dumb, that's it.

I don't think that should simply be asserted, but rather that this discussion is out of scope. After all, if we decide that a better overall solution includes changing the behavior of the Golem spell to "resummon" or even "do nothing," then this PR could end up changing.

Having said that, I would make the argument that regardless of whether the discussion is out of scope, there's no harm in accepting this PR as-is. This change is sensible based on the current behavior of the Golem spell, and it's so small that it would have virtually no impact on future solutions to the problems we're currently discussing.

**Comment by AJenbo** [comment]
Does this still have the hero go "I do not have enough mana" if low, but then cost no mana?

**Comment by qndel** [comment]
> Does this still have the hero go "I do not have enough mana" if low, but then cost no mana?

yup

**Comment by AJenbo** [comment]
Sounds a bit raw then.

**Comment by qndel** [comment]
> Sounds a bit raw then.

you are expecting too much from a simple `else` xD

**Comment by julealgon** [comment]
> you are expecting too much from a simple else xD

Well, you are fixing an issue and introducing a new problem.

**Comment by StephenCWills** [comment]
> you are expecting too much from a simple `else` xD

The fix may still be relatively small. FYI, I'm only making suggestions on things to try based on method names. I haven't actually tried changing any of these methods so I don't know what exactly is required.

The easiest thing I can think of would be to add a special case to `GetManaAmount()`.
https://github.com/diasurgical/devilutionX/blob/4f417d2fdcb2ed32f55d51bc25b5525b9c74f888/Source/spells.cpp#L15

Or, to avoid the mana cost in the menu changing, a special case in `CheckSpell()`.
https://github.com/diasurgical/devilutionX/blob/4f417d2fdcb2ed32f55d51bc25b5525b9c74f888/Source/spells.cpp#L183

**Comment by qndel** [comment]
> > you are expecting too much from a simple `else` xD
> 
> The fix may still be relatively small. FYI, I'm only making suggestions on things to try based on method names. I haven't actually tried changing any of these methods so I don't know what exactly is required.
> 
> The easiest thing I can think of would be to add a special case to `GetManaAmount()`.
> https://github.com/diasurgical/devilutionX/blob/4f417d2fdcb2ed32f55d51bc25b5525b9c74f888/Source/spells.cpp#L15
> 
> Or, to avoid the mana cost in the menu changing, a special case in `CheckSpell()`.
> https://github.com/diasurgical/devilutionX/blob/4f417d2fdcb2ed32f55d51bc25b5525b9c74f888/Source/spells.cpp#L183

golemspam

**Comment by FitzRoyX** [comment]
AJenbo how do you feel about categorizing this as a "click saver" simplification in the same spirit as Pepin auto-healing?
Golem casts always spawning a golem saves a click.

**Comment by qndel** [comment]
> AJenbo how do you feel about categorizing this as a "click saver" simplification in the same spirit as Pepin auto-healing?
> Golem casts always spawning a golem saves a click.

then it takes away the option to unsummon it

**Comment by julealgon** [comment]
@FitzRoyX it's not a click saver only. There are valid strategical reasons as to why you'd want to unsummon a golem.

**Comment by FitzRoyX** [comment]
> > AJenbo how do you feel about categorizing this as a "click saver" simplification in the same spirit as Pepin auto-healing?
> > Golem casts always spawning a golem saves a click.
> 
> then it takes away the option to unsummon it

We can just bind delete to unsummon, and make the graphical icon for it later. Besides, "kill this thing that can do nothing but help me" is a far less common desire than "make a new golem with fresh hp".

**Comment by AJenbo** [comment]
> > > AJenbo how do you feel about categorizing this as a "click saver" simplification in the same spirit as Pepin auto-healing?
> > > Golem casts always spawning a golem saves a click.
> > 
> > 
> > then it takes away the option to unsummon it
> 
> We can just bind delete to unsummon, and make the graphical icon for it later. Besides, "kill this thing that can do nothing but help me" is a far less common desire than "make a new golem with fresh hp".

I think you missed this comment: https://github.com/diasurgical/devilutionX/pull/1609#issuecomment-822786407

**Comment by FitzRoyX** [comment]
I'm not buying the argument that the golem creates more danger than it prevents. Even if it aggros something, it's hitting the golem first. You've got plenty of time to help him and go back to tetris. That's just how the golem is, why waste mana killing him off.

**Comment by StephenCWills** [comment]
> I'm not buying the argument that the golem creates more danger than it prevents. Even if it aggros something, it's hitting the golem first. You've got plenty of time to help him and go back to tetris. That's just how the golem is, why waste mana killing him off.

Because mana is cheap, and he's not worth the effort of babysitting his willful temperament when I can just nuke everything with Fireball at my leisure. Rather, you seem to think that resummoning him in one click is such a worthwhile change, but perhaps you should ask yourself three questions.

1. How often you resummon a Golem?
2. Why do you resummon him, and is it problematic to have to cast him twice to resummon?
3. Why did Beelzebub go to the trouble of implementing Unsummon if there's no value in having it?

In other words, perhaps a better use of your efforts would be to work on justifying the need for a click-save.

**Comment by FitzRoyX** [comment]
1. Vanilla stopped working on Windows for 10 years, so I can't comment on the golem's usefulness.
2. Golem is vital tankbait in bub. Even brief moments of being golemless can be deadly (haste and teleport bosses), so instant resummon is ideal. If it's anywhere near as useful in vanilla, I'd think you'd want minimal resummon time.
3. Never used it myself, but hotspells have to be swapped out and that's why I think it's better implemented as a hotkey.

**Comment by AJenbo** [comment]
2 and 3 can be dismissed because of 1

**Comment by qndel** [comment]
> 1. Vanilla stopped working on Windows for 10 years, so I can't comment on the golem's usefulness.
> 2. Golem is vital tankbait in bub. Even brief moments of being golemless can be deadly (haste and teleport bosses), so instant resummon is ideal. If it's anywhere near as useful in vanilla, I'd think you'd want minimal resummon time.
> 3. Never used it myself, but hotspells have to be swapped out and that's why I think it's better implemented as a hotkey.

You seem to be the only one having problems, I'm running vanilla on windows 10 just fine, if you haven't noticed, this is devilutionx, not belzebub, stop comparing everything to belzebub, did you play vanilla at all? >.>

**Comment by FitzRoyX** [comment]
> You seem to be the only one having problems, I'm running vanilla on windows 10 just fine

Congrats, I get no video just sound. Eventually a third party made a ddraw dll and that kinda works.

> , if you haven't noticed, this is devilutionx, not belzebub

And if you haven't noticed, this is devilutionx, not devilution. No one wants to click on Pepin heal, no one wants to click per attack, no one wants to cast golem twice to respawn it. It's not my fault bub is so easily referenced in every issue that vanilla had. Someone literally just came into discord and asked about town automap and sprite limits. Guess what? Bub has town automap and no sprite limits. You are going to see endless crossover here and your solutions will very often match his. There's no need to ignore that, I think it's a great reference.

**Comment by nickgavran** [comment]
Dude. This is getting ridiculous at this point.

I have a fully functioning vanilla and hellfire, on Windows 10. No extra files or workarounds required, it worked straight out of the box. Sorry it doesn't work for you personally, but that is not the norm.

Secondly, belzebub shares absolutely nothing with diablo codewise. It's a completely different engine written from scratch, the only similarity they have is the fact that belzebub uses the diablo assets for graphics.

If you think the two are still comparable, I'll start comparing pacman to crysis for future discussions, since both games require you to kill enemies, that's a fair assumption to make right?

**Comment by StephenCWills** [comment]
@FitzRoyX Based on my personal experience, Golem is actually mostly useless in vanilla. He does become somewhat more useful in Hell difficulty, but not as tankbait. Quite the opposite, actually. He speeds up runs against the triple-immune Soul Burners (witches) and Advocates (mages). Seeing as how I don't play on Hell difficulty all that often, I'll have to defer to text from Flux's mage guide to help explain...

https://diablo2.diablowiki.net/Flux%27s_Expert_Mage_Guide#Tactics
> Once I reach Laz's room and then for lvl 16, (since there are advocates in with Laz, it's unofficially the start of lvl 16.) I switch Chain Lightning for Golem. Golem is very mana-expensive, but it is indispensable for dealing with triple immunes. It's great for advocates, and still useful for Soul Burners, but must be used more judiciously, as any type of witch will target a golem from any distance, and unless you have him virtually in a corner, he's going to be bloodstar bait and then gravel in just a few seconds.
> 
> Recommendation is to still use him a lot on lvl 15 if you get Soul Burners, but just try to make sure you have all the live witches in the area on stone before you pop him out, or he's not going to last long enough to help much. Soul Burners are probably not as bad as Advocates, due to the much lower damage of their bloodstars compared to the Advocates fireballs, and their less unpredictable escape patterns, but they have many more hps, so a golem's help in killing them is really a must if you don't want to spend an hour on lvl 15.

**Comment by julealgon** [comment]
> Vanilla stopped working on Windows for 10 years

I went through Win98, Vista, Windows 7, Windows 8, Windows 8.1 and now Windows 10 without any issues running vanilla Diablo beside the occasional color shifts that are fixed by closing the explorer process (very known issue/fix).

🤷‍♂️

**Comment by FitzRoyX** [comment]
Dunno what to tell you, it's a thing, and it was harder to find a solution in 2010 when I first gave up
https://www.lurkerlounge.com/forums/thread-13200.html
https://www.gog.com/forum/diablo/game_crashes_after_showing_blizzard_logo
https://www.youtube.com/watch?v=_mKxU_PNwX4
https://www.reddit.com/r/Diablo/comments/3v13fw/diablo_1_on_windows_10_any_luck/
http://www.d3andre.com/en/diablo-1-how-to-run-on-windows-10.html
https://www.reddit.com/r/Diablo/comments/9w72uu/struggling_to_get_diablo_1_to_run_on_windows_10/
https://www.reddit.com/r/Diablo1/comments/3h8v62/fix_for_windows_10_crashes/
https://www.ttlg.com/forums/showthread.php?t=138366
https://forum.winehq.org/viewtopic.php?t=28485
https://gaming.stackexchange.com/questions/35210/playing-diablo-1-on-windows-7
https://www.diablofans.com/forums/diablo-forums/diablo-legacy-forums/diablo-i-hellfire/70404-cant-play-diablo-1-becoming-crazy-help-me

**Comment by qndel** [comment]
could go well with the ability to remove manashield for free

**Comment by StephenCWills** [comment]
This PR has no conflicts, but it's referencing old variables so it needs to be rebased and fixed before merging. It's certainly possible to repeatedly unsummon your golem by spamming the spell like in @qndel's gif, but that doesn't seem to have any negative side effects, and it's certainly not an issue caused by this PR.

**Comment by qndel** [comment]
> This PR has no conflicts, but it's referencing old variables so it needs to be rebased and fixed before merging. It's certainly possible to repeatedly unsummon your golem by spamming the spell like in @qndel's gif, but that doesn't seem to have any negative side effects, and it's certainly not an issue caused by this PR.

well, earlier it would've cost you mana to do so = only "special" people would do it, and since it's free now, someone could do it and it's ugly 😅
